src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,444 @@
+        val transition = IceProcessingStateTransition(oldState, newState)
+
+        logger.info("ICE state changed old=$oldState new=$newState")
+
+        when {
+            transition.completed() -> {
+                if (iceConnected.compareAndSet(false, true)) {
+                    eventHandler?.connected()
+                }
+            }
+            transition.failed() -> {
+                if (iceFailed.compareAndSet(false, true)) {
+                    eventHandler?.failed()
+                }
+            }
+        }
+    }
+
+    private fun iceStreamPairChanged(ev: PropertyChangeEvent) {
+        if (IceMediaStream.PROPERTY_PAIR_CONSENT_FRESHNESS_CHANGED == ev.propertyName) {
+            /* TODO: Currently ice4j only triggers this event for the selected
+             * pair, but should we double-check the pair anyway?
+             */
+            val time = Instant.ofEpochMilli(ev.newValue as Long)
+            eventHandler?.consentUpdated(time)
+        }
+    }
+
+    companion object {
+        fun appendHarvesters(iceAgent: Agent) {
+            Harvesters.initializeStaticConfiguration()
+            Harvesters.tcpHarvester?.let {
+                iceAgent.addCandidateHarvester(it)
+            }
+            Harvesters.singlePortHarvesters?.forEach(iceAgent::addCandidateHarvester)
+
+            iceAgent.setUseHostHarvester(false)
+        }
+    }
+
+    private data class PacketStats(
+        var numPacketsReceived: Int = 0,
+        var numIncomingPacketsDroppedNoHandler: Int = 0,
+        var numPacketsSent: Int = 0
+    ) {
+        fun toJson(): OrderedJsonObject = OrderedJsonObject().apply {
+            put("num_packets_received", numPacketsReceived)
+            put("num_incoming_packets_dropped_no_handler", numIncomingPacketsDroppedNoHandler)
+            put("num_packets_sent", numPacketsSent)
+        }
+    }
+
+    interface IncomingDataHandler {
+        /**
+         * Notify the handler that data was received (contained
+         * within [data] at [offset] with [length]) at [receivedTime]
+         */
+        fun dataReceived(data: ByteArray, offset: Int, length: Int, receivedTime: Instant)
+    }
+
+    interface EventHandler {
+        /**
+         * Notify the event handler that ICE connected successfully
+         */
+        fun connected()
+        /**
+         * Notify the event handler that ICE failed to connect
+         */
+        fun failed()
+        /**
+         * Notify the event handler that ICE consent was updated
+         */
+        fun consentUpdated(time: Instant)
+    }
+}
+
+/**
+ * Models a transition from one ICE state to another and provides convenience
+ * functions to test the transition.
+ */
+private data class IceProcessingStateTransition(
+    val oldState: IceProcessingState,
+    val newState: IceProcessingState
+) {
+    // We should be using newState.isEstablished() here, but we see
+    // transitions from RUNNING to COMPLETED, which should not happen and

```

Comment:
I believe `COMPLETED` should say `TERMINATED` (I know the original has the same mistake).

Reply:
If you fix this, maybe add a note that this happens when the Agent is freed prior to being started

Reply:
Also, the comment should apply to `failed` below. Looking at the git history it was me who messed it up originally...

Reply:
Sure, will tweak the comment.